### PR TITLE
Add an example of few-shot memory bank model with MultiModal Feature Extraction

### PR DESCRIPTION
**Pull request by Linuxdex**
This is an example which provides a simple and clear way of implementing a memory-bank-powered few-shot learning model with AutoGluon MultiModal according to [Tip-adapter](https://arxiv.org/abs/2111.03930). 

The idea is to store `<feature, label>` pairs from the training data in a key-value memory bank. In the prediction phase, we compare the similarity between the test image features and the memory bank keys, and aggregate the prediction logits. The logits obtained via feature-similarity is combined with logits obtained from a classification model that directly predicts the label from the features.

Experiments show that adding a memory-bank can improve the performance of image, text and image-text classification in the few-shot learning scenario.

### CODE REVIEW

File: examples/automm/cache_adapter/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Use MultiModal Feature Extraction to Create a Few-shot Cache Adapter Model
+
+[cache_adapter.py](./cache_adapter.py): This example provides a simple and clear way to implement a few-shot cache adapter model with AutoGluon MultiModal Feature Extraction according to [Tip-adapter](https://github.com/gaopengcuhk/Tip-Adapter) [1]. 
+
+### 1. Run Example
+
+Before running the example, the image datasets need to be prepared with additional code. We follow the few-shot dataset design in [CoOP](https://github.com/KaiyangZhou/CoOp). Copy folder `configs` and `datasets` from the repository and rename `datasets` to `imagedatasets` to avoid conflict with `datasets` in Hugging Face. You can download the datasets and the splits under the [direction]().
```
Comment (bryanyzhu) [inline examples/automm/cache_adapter/README.md:7]: It seems the direction in the end of the sentence is not linked yet?

File: examples/automm/cache_adapter/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    python cache_adapter.py --type clip --dataset food101 --shots 16
+
+- `type` is the type of few-shot learning. You choose from `clip`, `text`, `image` for different backbone methods.
+- `backbone` is the backbone model of MultiModal Predictor.
+- `data_path` is the path for image dataset.
+- `dataset` is the name of dataset. Support image datasets in COOP and text datasets in Hugging Face.
+- `column_names` is the names of the data columns.
+- `label_column` is the name of the label column.
+- `shots` is the shots for each class in training set.
+- `aug_epochs` is the epochs to create the cache.
```
Comment (bryanyzhu) [inline examples/automm/cache_adapter/README.md:22]: Actually I think we need a bit more context for what is cache adapter. Some people might know adapter, but not many people know what is cache adapter and what does cache stand for. So in this case, why we need the epochs to create the cache?

File: examples/automm/cache_adapter/README.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# Use MultiModal Feature Extraction to Create a Few-shot Cache Adapter Model
+
+[cache_adapter.py](./cache_adapter.py): This example provides a simple and clear way to implement a few-shot cache adapter model with AutoGluon MultiModal Feature Extraction according to [Tip-adapter](https://github.com/gaopengcuhk/Tip-Adapter) [1]. 
+
+### 1. Run Example
+
+Before running the example, the image datasets need to be prepared with additional code. We follow the few-shot dataset design in [CoOP](https://github.com/KaiyangZhou/CoOp). Copy folder `configs` and `datasets` from the repository and rename `datasets` to `imagedatasets` to avoid conflict with `datasets` in Hugging Face. You can download the datasets and the splits under the [direction]().
+
+For text datasets, datasets in Hugging Face are supported. We recommend datasets under SetFit for few-shot learning. Remember to change the column names in args if the data column is not named `text` or have multi-text information.
```
Comment (bryanyzhu) [inline examples/automm/cache_adapter/README.md:9]: Can you help to explain what is SetFit?

File: examples/automm/cache_adapter/cache_adapter.py [hunk 4]
```diff
@@ -0,0 +1,465 @@
+"""
+    The derivative application of Tip-Adapter (https://arxiv.org/pdf/2111.03930.pdf).
+    Refer to https://github.com/gaopengcuhk/Tip-Adapter
```
Comment (bryanyzhu) [inline examples/automm/cache_adapter/cache_adapter.py:3]: It is better to include the license information of the repo as well

File: examples/automm/cache_adapter/cache_adapter.py [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+        """
+        Generate logits with cache adapter based on pure_logits and cache output.
+
+        Parameters
+        ----------
+        affinity
+            The result of cache similarity. It is based on cosine similarity or a projector initialized with cache_keys.
+        pure_logits
+            The predict proba of the classifier.
+        alpha
```
Comment (bryanyzhu) [inline examples/automm/cache_adapter/cache_adapter.py:108]: Both alpha and beta are shown as hyperparameters of cache model. What do they represent? Do they have default values? Please add more details to the documentation, thank you.

**Review by bryanyzhu — COMMENTED** [review]
Thanks for the PR, performance looks good.

**Comment by sxjscience** [comment]
@Linuxdex , Let's add more description of the memory cache structure.

File: examples/automm/cache_adapter/README.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
+# Use MultiModal Feature Extraction to Create a Few-shot Cache Adapter Model
```
Comment (sxjscience) [inline examples/automm/cache_adapter/README.md:1]: I also feel that cache adapter is not a good name.

File: examples/automm/cache_adapter/README.md (hunk 3 above)
Comment (Linuxdex) [reply examples/automm/cache_adapter/README.md:9]: It is a few-shot method of text classification (https://arxiv.org/pdf/2209.11055.pdf). Here, I think using the datasets under SetFit is more suitable for few-shot tasks.

File: examples/automm/cache_adapter/README.md (hunk 2 above)
Comment (Linuxdex) [reply examples/automm/cache_adapter/README.md:22]: For preprocessing with randomness, the mean of multi-generated caches will gain robustness. However, it still performs good without this process. So I set its default=1.

**Comment by sxjscience** [comment]
@Linuxdex You may also compare with the baseline of feature extraction + SVM, which has been added in https://github.com/autogluon/autogluon/pull/2850

**Review by sxjscience — APPROVED** [review]
LGTM. We can revise the textual descriptions in another PR.

File: examples/automm/memory_bank/README.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+Memory bank which is the derivative application of Tip-Adapter obtains diversified multi-modal features through MultiModal Feature Extraction. In this example, we first trained a linear classifier based on multi-modal features to generate baseline accuracy. Then, the similarity result between features and memory bank is introduced to baseline predict probability. Finally, an additional linear adapter which is initialized with memory bank is trained to help few-shot classification.
+
+Hyper-parameters `alpha` and `beta` which adjust the memory bank are modified through grid search on validation set to attain the superior performance.
```
Comment (sxjscience) [inline examples/automm/memory_bank/README.md:39]: Same here. Need to revise the paragraph.

File: examples/automm/memory_bank/memory_bank.py [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+            The hyper-parameters of memory bank model.
+        beta
+            The hyper-parameters of memory bank model.
+        
+        Return
+        ------
+        The predict probability of the feature.
+            - "pure_logits"
+                The predict probability of classifier.
+            - "logits_with_adapter"
```
Comment (sxjscience) [inline examples/automm/memory_bank/memory_bank.py:179]: Consider to rename to `adapted_logits`.

File: examples/automm/memory_bank/memory_bank.py [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+                The predict probability composed of classifier and memory bank similarity result.
+            - "logits_with_finetuned_adapter"
```
Comment (sxjscience) [inline examples/automm/memory_bank/memory_bank.py:181]: Consider to rename to `adapted_logits_with_finetuning`

File: examples/automm/memory_bank/memory_bank.py [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+        bank_keys
+            The content of bank composed of features in the training set.
+        bank_labels 
+            The labels of corresponding bank_keys.
+        hidden_size
+            The size of features.
+        num_classes
+            The classes of the dataset.
+        clip_weights
+            The encoded label-text prompts which is only used in type "clip".
```
Comment (sxjscience) [inline examples/automm/memory_bank/memory_bank.py:91]: We can change it to `clip embedding of the semantic text that describes the labels`.

File: examples/automm/memory_bank/utils.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    for per_name in column_names:
+        per_features = torch.tensor(predictor.extract_embedding({per_name: data[per_name]})[per_name]).cuda()
+        features.append(per_features)
+    features = torch.stack(features, dim=1).cuda()
+    features = features.mean(dim=1)
+    features /= features.norm(dim=-1, keepdim=True)
+    return features
+
+
+def generate_clip_weights(args, classnames, template, predictor):
```
Comment (sxjscience) [inline examples/automm/memory_bank/utils.py:68]: We can consider to just introduce the variable called `semantic_label_embedding` and compare the similarity between `semantic_label_embedding` and the embeddings of the labels stored in the memory bank.

File: examples/automm/memory_bank/memory_bank.py [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+        if clip_weights is None:
+            if model_head_type == "SVM":
+                self.model_head = make_pipeline(StandardScaler(), SVC(gamma="auto", probability=True))
+                self.model_head.fit(bank_keys.t().cpu(), bank_labels.cpu())
+            else:
+               self.model_head = nn.Linear(hidden_size, num_classes, bias=True) if clip_weights is None else None
+        else:
+            self.model_head = None
+
+    def get_adapter_logits(self, affinity, pure_logits, alpha, beta):
```
Comment (sxjscience) [inline examples/automm/memory_bank/memory_bank.py:113]: We can call it 

```python
def adapt_logits(...)

```
